x/supply/keeper/keeper.go

```diff
@@ -0,0 +1,84 @@
+package keeper
+
+import (
+	"github.com/cosmos/cosmos-sdk/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	authkeeper "github.com/cosmos/cosmos-sdk/x/auth/keeper"
+	"github.com/cosmos/cosmos-sdk/x/auth/vesting/exported"
+	bankkeeper "github.com/cosmos/cosmos-sdk/x/bank/keeper"
+	distributionkeeper "github.com/cosmos/cosmos-sdk/x/distribution/keeper"
+	"github.com/desmos-labs/desmos/v3/x/supply/types"
+	"github.com/tendermint/tendermint/libs/log"
+)
+
+type Keeper struct {
+	cdc codec.BinaryCodec
+	ak  authkeeper.AccountKeeper
+	bk  bankkeeper.Keeper
+	dk  distributionkeeper.Keeper
+}
+
+// NewKeeper creates new instances of the supply keeper
+func NewKeeper(cdc codec.BinaryCodec,
+	ak authkeeper.AccountKeeper, bk bankkeeper.Keeper, dk distributionkeeper.Keeper) Keeper {
+	return Keeper{
+		cdc: cdc,
+		ak:  ak,
+		bk:  bk,
+		dk:  dk,
+	}
+}
+
+func (k Keeper) Logger(ctx sdk.Context) log.Logger {
+	return ctx.Logger().With("module", "x/"+types.ModuleName)
+}
+
+// GetConvertedTotalSupply returns the total supply converted by dividing it for the given divider factor
+func (k Keeper) GetConvertedTotalSupply(ctx sdk.Context, coinDenom string, divider sdk.Int) sdk.Int {
+	totalSupply := k.bk.GetSupply(ctx, coinDenom)
+	return totalSupply.Amount.Quo(divider)
+}
+
+// CalculateCirculatingSupply calculates the current circulating supply by:
+// 1. Getting the total supply
+// 2. Subtract the community pool amount from it
+// 3. Subtract the vesting accounts locked tokens from it
+// It then returns it converted by diving it with the divider factor
+func (k Keeper) CalculateCirculatingSupply(ctx sdk.Context, coinDenom string, divider sdk.Int) sdk.Int {
+	var circulatingSupply sdk.Int
+
+	// Get total supply
+	totalSupply := k.bk.GetSupply(ctx, coinDenom)
+
+	// Get the community pool denom amount
+	communityPoolDenomAmount := k.dk.GetFeePoolCommunityCoins(ctx).AmountOf(coinDenom)
+
+	// Subtract community pool amount from the total supply
+	circulatingSupply = totalSupply.Amount.Sub(communityPoolDenomAmount.RoundInt())
+
+	// Subtract all vesting account locked tokens amount from the circulating supply
+	accounts := k.ak.GetAllAccounts(ctx)
```

Comment:
I am curious about what if there are millions or billions accounts. Would it load all the accounts into the memory then crash the server? If yes, I think we should use `ak.IterateAccounts` instead of `GetAllAccounts`.

Reply:
You're right! Thanks for pointing this out @dadamu 👍